### PR TITLE
fix fromHoconString producing exception on parse error (#325)

### PR DESCRIPTION
**Pull request by fsarradin**
I have added a call-by-name for `fromTypesafeConfig` and a unit test to check that we get a Left value instead of an exception.

See issue #325

### CODE REVIEW

File: typesafe/src/test/scala/zio/config/typesafe/TypesafeConfigErrorsSpec.scala [hunk 1]
```diff
@@ -101,6 +101,17 @@ object TypesafeConfigErrorsSpec extends DefaultRunnableSpec {
...
+    test("A variant error case with a not well-formed typesafe HOCON config") {
+      val hocconStringWithParseError =
+        s"""
+         account {
+        """
+
+      val notWellFormedConfigResult = TypesafeConfigSource.fromHoconString(hocconStringWithParseError)
+      val notWellFormedConfigExpect = Left("String: 3: expecting a close parentheses ')' here, not: end of file")
+
+      assert(notWellFormedConfigResult)(equalTo(notWellFormedConfigExpect))
```
Comment (ahoy-jon) [inline typesafe/src/test/scala/zio/config/typesafe/TypesafeConfigErrorsSpec.scala:113]: This test is very specific. I would go for something like:
```scala
assert(notWellFormedConfigResult.isLeft)(Assertion.isTrue)
```

Comment (fsarradin) [reply typesafe/src/test/scala/zio/config/typesafe/TypesafeConfigErrorsSpec.scala]: Agreed. Modification done.

**Comment by ahoy-jon** [comment]
Indeed we* are using this kind of technique when we are wrapping impure type: 
https://github.com/univalence/zio-spark/blob/master/src/main/scala/zio/spark/wrap/Wrap.scala#L12-L44
(With a lot of dependent typing here)

However, the technique of creation would be something like :
````
object ZTypesafeConfig {
   def fromString(string: String): Task[ZTypesafeConfig]  =  ??? 
   def default: UIO[ZTypesafeConfig]  =   ???
   // In future lot many others may be

}
```
Because the config relies on mutable elements in the JVM environment.

* Working in the same organization, @univalence

**Comment by ahoy-jon** [comment]
@afsalthaj That being said, I don't understand why this PR is not merged, and I see a pattern emerging in ZIO contributions (we (try to) contribute to ZIO).

1. There is an issue with a **harmful** bug, related to the initial design
2. There is a minimal Pull Request
3. The Pull Request comes with a valid test

Yet it is not merged, because the PR is turned as a vehicle to change or to discuss the initial design (with a new contributor).

Could you: 
- create a new issue to change the design?
- merge this PR?

I am not trying to be mean, just exaggerating a bit! 😀

**Comment by afsalthaj** [comment]
The change I proposed doesn’t need any redesign. It doesn’t need to return a `Task` in smart constructor space.

It’s a small change. Merging this one, once the build is over.

**Comment by afsalthaj** [comment]
@ahoy-jon  Thanks for your contribution

**Pull request merged**